Microsoft Docs samples. (#10)
Azure-Samples/azure-monitor-opencensus-python: docs_samples/01_python_app_insights_overview.md

  - ms-python.python
  - ms-azuretools.vscode-azurefunctions

## Check your Python Version

- In Visual Studio Code, open a Terminal window, run the following command:

```powershell
python.exe --version
```

- For Python, make sure you see version 3.9.x.  If you do not, do the following:
  - Download Python and change your path variables to point to the proper version
  - Restart Visual Studio Code
- Open the command palette by selecting **View->Command Pallet**
- Run the `Python: Select interperter` command
- Select the `Python 3.9.x` version

## Create a Python environment

- Run the following command to create an environment where you can load all the dependencies:

```powershell
py -3 -m venv .venv
.venv\scripts\activate
```

- When prompted, select **yes**
- Ensure that you select the new environment in the interpertor selection otherwise the python commands you run later may not map to the proper python version.

## Install OpenCensus

[OpenCensus](https://opencensus.io/) is a set of open source libraries hosted on [GitHub](https://github.com/census-instrumentation) that are used to capture trace and metric data from applications.

> Metrics can be things like latency, HTTP request and reponse lengths whereas traces are how a request moves through your application layers and other services. Graphs can be built to show how an application or service is performing based on these data points.

Reviewer: ```suggestion
Metrics can be things like latency, HTTP request and reponse lengths whereas distributed tracing shows how a request moves through your application layers and other services. Log data represents user generated logs that can be enriched with context and custom dimensions. Graphs can be built to show how an application or service is performing based on these data points.
```

To install OpenCensus:

- Open a terminal window, run the following:

```powershell
python -m pip install --upgrade pip

python -m pip install opencensus
```

## Install Azure exporter

> OpenCensus needs an exporter to know how to send the log and metric data to a specific backend.  Azure provides an exporter for Azure Monitor and Log Analytics.

Reviewer: ```suggestion
OpenCensus needs an exporter to know how to send the log and metric data to a specific backend.  Azure provides an exporter for Azure Monitor.
```

> OpenCensus needs an exporter to know how to send the log and metric data to a specific backend.  Azure provides an exporter for Azure Monitor and Log Analytics.

Reviewer: ```suggestion
OpenCensus needs an exporter to know how to send the log, metric and tracing data to a specific backend.  Azure provides an exporter for Azure Monitor and Log Analytics.
```

To install the Azure monitor exporter package:

- Open a terminal window, run the following:

```powershell
python -m pip install opencensus-ext-logging
python -m pip install opencensus-ext-azure
python -m pip install opencensus-extension-azure-functions
python -m pip install opencensus-ext-requests
python -m pip install pyodbc
python -m pip install psutil
```

## Configure middleware to track requests

- Browse to the Azure Portal
- Select the Application Insights **python-appinsights-SUFFIX** resource
- On the **Overview** page, copy the connection string

> - Open the `.\docs_sampes\.env` file notice that as part of the setup the values have been copied into the environment file

Reviewer: ```suggestion
- Open the `.\docs_samples\.env` file notice that as part of the setup the values have been copied into the environment file
```

> 

Reviewer: I would include here to verify that the connection strings match from the resource and the settings file.

> # Getting Stared with App Insights for Python

Reviewer: ```suggestion
# Getting Started with App Insights for Python via OpenCensus Azure monitor exporters
```

> This procedure configures your Python application to send telemtery tot he Application Insights feature of the Azure Monitor service.  It works with any Python application on-premises or in the cloud.

Reviewer: ```suggestion
This procedure configures your Python application to send telemtery to the Application Insights feature of the Azure Monitor service.  It works with any Python application on-premises or in the cloud.
```

## Prerequisites

> To work with App Insights for Python, you need to:

Reviewer: ```suggestion
To get started with the following samples, you need to:
```

> - Install Python (version 3.9.x).

Reviewer: Do these examples only work on v3.9.x?

Author: They should run on 3.7.x or higher.  I can redo all the steps with 3.7 to verify if you think its worth it to do so?

Reviewer: No need, I would just add an excerpt stating that they have been CONFIRMED to work with Python 3.9 and not restrict the user to a certain version.

- Install the latest version of Visual Studio Code with the following extensions.
  - ms-python.python
  - ms-azuretools.vscode-azurefunctions
- Install [Azure Functions Core Tools](https://docs.microsoft.com/en-us/azure/azure-functions/functions-run-local?tabs=v4%2Cwindows%2Ccsharp%2Cportal%2Cbash#install-the-azure-functions-core-tools)
- Create a [free Azure account](https://azure.microsoft.com/free/) if you don't already have an Azure subscription.
- [Azure Storage Emulator](https://docs.microsoft.com/en-us/azure/storage/common/storage-use-emulator) (for local testing)

## Deploy the environment

- Open a Windows PowerShell window
- Run the following:

```Powershell
cd $home;

rm -r python-appinsights -f

> git clone https://github.com/solliancenet/docs-python docs-python

Reviewer: Link to this repo instead?

```

- Open the Azure Portal
- Create a new resource group

> - Run the `./Artifacts/template.json` file

Author: great point.  Updated to have the full steps.

  - Restart Visual Studio Code
- Open the command palette by selecting **View->Command Pallet**
- Run the `Python: Select interperter` command
- Select the `Python 3.9.x` version

## Create a Python environment

- Run the following command to create an environment where you can load all the dependencies:

```powershell
py -3 -m venv .venv
.venv\scripts\activate
```

- When prompted, select **yes**
- Ensure that you select the new environment in the interpertor selection

## Install OpenCensus

[OpenCensus](https://opencensus.io/) is a set of open source libraries hosted on [GitHub](https://github.com/census-instrumentation) that are used to capture trace and metric data from applications.

Metrics can be things like latency, HTTP request and reponse lengths whereas traces are how a request moves through your application layers and other services. Graphs can be built to show how an application or service is performing based on these data points.

To install OpenCensus:

- Open a terminal window, run the following:

```powershell
pip install --upgrade pip

python -m pip install opencensus
```

## Install Azure exporter

> OpenCensus needs an exporter to know how to send the log and metric data to the target.  Azure provides an exporter for Azure Monitor.

Reviewer: ```suggestion
OpenCensus needs an exporter to know how to send the log and metric data to a specific backend.  Azure provides an exporter for Azure Monitor.
```

> To install the Azure Exporter:

Reviewer: ```suggestion
To install the Azure monitor exporter package:
```

- Open a terminal window, run the following:

```powershell
python -m pip install opencensus-ext-logging
python -m pip install opencensus-ext-azure
python -m pip install opencensus-extension-azure-functions
python -m pip install opencensus-ext-requests
python -m pip install pyodbc
python -m pip install psutil

```

## Configure middleware to track requests

- Browse to the Azure Portal
- Select your Azure Monitor resource
- On the **Overview** page, copy the connection string

> - Open the `.env` file and paste it into the `APPLICATIONINSIGHTS_CONNECTION_STRING` value

Reviewer: These steps that involve a specific file are currently not clear to me. Are they generated  from``./Artifacts/template.json` above?

Author: I updated this to be automated now as part of a setup.ps1 file.  People won't really need to do anything except press F5 to run things!

## Create a simple Python App

- Browse to the Azure Portal.
- Select your Azure Monitor resource.
- On the **Overview** page, copy the connection string and save it for later.
- Create a new python file called `app.py`.
- Copy the following into it, be sure to replace your Application Insights connection string:

```python
import logging

> from opencensus.ext.azure.log_exporter import AzureLogHandler

Reviewer: What is the purpose of the document? This sample uses an example that is already existing in [examples](https://github.com/census-instrumentation/opencensus-python/blob/master/contrib/opencensus-ext-azure/examples/logs/simple.py).

logger = logging.getLogger(__name__)

logger.addHandler(AzureLogHandler(connection_string='InstrumentationKey=<your-instrumentation_key-here>'))

logger.warning('Hello, World!')
```

- Press **F5** to run the file, select **Python file** in the debug configuration window

> - If you get an error about `psutil._psutil_windows` do the following:

Reviewer: How does this error occur? Isn't this just a simple example for logging?

```

- You should see the following:

  ![The query is displayed with one result from the above program.](./media/python_simple_app_trace.png "Review the results of the query.")

## Sending trace data

- Open the `./SimpleApps/trace.py` file
- Press **F5** to run the file
- Switch to the Azure Portal
- Browse to your lab resource group
- Browse to the `python-appinsights-SUFFIX` application insights resource and select it
- Under **Monitoring**, select **Logs**
- Run the following Kusto query to see the `traces` sent to application insights:

```kql
traces
| sort by timestamp
| where cloud_RoleName == "trace.py"
```

- You should see your trace in the **message** column:

  ![The query is displayed with one result from the above program.](./media/python_simple_trace_trace.png "Review the results of the query.")

- For your trace item notice that some of the columns are empty, but others (such as `cloud_RoleName`, `cloud_RoleInstance` and `client_*`) are populated based on the client information from the SDK.

## Logger methods

When working with the [Python logging framework](https://docs.python.org/3/library/logging.html), you should take note of the following methods:

> 

Reviewer: What is this section for? What does it mean for the user to "take note"?

Author: removed it...was going to put more content in it but decided to just delete it.

- addHandler
- warning
- exception
- setLevel
- info
- debug
- critical

> ## Capture exceptions

Reviewer: ```suggestion
## Capture exceptions and custom dimensions
```

- Switch back to your Visual Studio Code window
- Select the `./SimpleApps/exception.py` file
- Press **F5** to run the file, select **Python file** in the debug configuration window
- Switch to the Azure Portal
- Browse to your lab resource group
- Browse to the `python-appinsights-SUFFIX` application insights resource and select it
- Under **Monitoring**, select **Logs**
- Run the following Kusto query to see the `exception` sent to application insights:

```kql
exceptions
| sort by timestamp
| where cloud_RoleName == "exception.py"
```

- Expand your exception item, notice the `details` column has the exception details you would expect, but also expand the `customDimensions` column to review the custom data that was sent to the application insights table.

  ![The query is displayed with one result from the above program.](./media/python_simple_exception_custom.png "Review the results of the query.")

> ## Utilize custom events and metrics

Reviewer: Curious as to how the below function apps has to with "Utilize custom events and metrics"? Opencensus isn't even being used in those examples.

Author: I changed the heading of this.  Also moved the function stuff to dependencies as that was what I intended to show originally.

- Switch to Visual Studio Code

> - Open a terminal window, run the following commands, be sure to replace the `SUFFIX`:

Author: Added a little extra instructions.

```powershell
cd Functions

az login

> func azure functionapp publish python-appinsights-SUFFIX   

Reviewer: Probably would be better to separate out function app examples into its own file `04_python_app_functions.md` and show examples of how to use the functions [extension](https://github.com/census-ecosystem/opencensus-python-extensions-azure/tree/main/extensions/functions#opencensus-python-azure-functions-extension)
